docs/developer/architecture/development-plugin-saved-objects.asciidoc

```diff
@@ -202,4 +202,226 @@ up to date. If a visualization `id` was directly stored in
+persists the state of the underlying charts and allows those charts to define
+migrations for that state.
+
+The example code here is length, but necessary to demonstrate how this different
+pieces fit together.
+
+src/plugins/dashboard/server/saved_objects/dashboard.ts
+[source,typescript]
+----
+type ChartMigrationFn<ChartState extends object = object> = (chartDoc: object) => ChartState;
+
+interface Chart<ChartState extends object = object> {
+  type: string;
+  render(chart: ChartState, element: HTMLElement): void;
+  create(): ChartState;
+  getMigrations(): Record<string, ChartMigrationFn<ChartState>>;
+}
+
+class ChartRegistryPlugin {
+  private readonly charts = new Map<string, Chart>();
+  public setup() {
+    return {
+      register<ChartState extends object>(chart: Chart<ChartState>) {
+        charts.set(chart.id, chart);
+      },
+
+      /** Returns migrations by version that can handle migrating a chart of any type */
+      getMigrations() {
+        // Here we rollup the migrations from each individual chart implementation to create a single migration function
+        // for each version that will the chart state migration if the input chart is of the same type.
+        const chartMigrations = this.charts.reduce((migrations, chart) => {
+          for (const [version, chartMigration] of Object.entries(chart.getMigrations)) {
+            const existingMigration = migrations[version] ?? (input: any) => input;
+            migrations[version] = (chartDoc) => {
+              if (chartDoc.type === chart.type) {
+                chartDoc = chartMigration(chartDoc);
+              }
+
+              return existingMigration(chartDoc);
+            }
+          }
+        }, {} as Record<string, ChartMigrationFn>);
+        
+        return {
+          '1.1.0': (oldChart) => {
+            return {
+              ...oldChart,
+              // Apply the migrations on the chart state if and only if any charts define migrations for this version.
+              // It's important that the registry items can only access the `state` part of the complete chart object.
+              // This is the part of the document that the chart implementation functionally "owns".
+              state: chartMigrations['1.1.0'] ? chartMigrations['1.1.0'](oldChart.chartState) : oldChart.chartState
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+interface BarChartState100 {
+  xAxis: string;
+  yAxis: string;
+  indexPattern: string;
+}
+
+interface BarChartState110 {
+  xAxis: string;
+  yAxis: string;
+  dataSource: {
+    indexPattern: string;
+  }
+}
+
+class BarChartPlugin {
+  public setup(core, plugins) {
+    plugins.charts.register<BarChartState100>({
+      type: 'bar-chart',
+      render() { ... },
+      create() { return { xAxis: 'foo', yAxis: 'bar', dataSource: { indexPattern: 'baz' } }},
+      getMigrations() {
+        return {
+          '1.1.0': (oldChartState: BarChartState100): BarChartState110 => ({
```

Comment:
I don't think the types in this example will work:

https://www.typescriptlang.org/play?#code/LAKALgngDgpgBAYQBYEMBOYCyBLA5mlMbAewDsAxUgHmXTAGUxD4YAPMGUgEwGc5iARgCsYAYzBwAvP2FiwAPilwAFKNQYAIsVEAuGSPEBKKYtoZGzANyhQ2UhzQAzFKPhmwNdQyYc4bDtx8ggYS0sFyigDeoHBwkLB6PGBodrjWILFonFwwaKpeeu4WHAA0fgA2MAC2nGB6ABIAKpgAMgCilTX2hnoAbsTYXOmxolnMyj2IXsUww3C4MFh4BERkPBN6AEpixGhcVEkppLhl7jj4hCQU1EU+MPLy6QC+NiCiaxJqdHzSpDAA7nBMCgoAdkqlTl55BN0qBHABXUjiK5wLK4bBJXKeOgzPzsbJBWTiaFfDCFLzY8x3eTGaIZOCksA8AB0PEW+TozPiMDKjMMz1edgczlccAAQuhbswAIwABllcDpsVYAEFWBjEuDjnMIGqNXBDqk5nYcqwAAqEBykTVHNKgF4gIW5EXwCVoKUcaVyxUxOCq9U8G1G326gNB7W+riEFD0YjwtCuPRK2JwE1sC1gK3hu30h0O0BojEOKhuj0wOWy6HJ7l6ADkAnQAFpGbWSr6stxchNFXAnm36aMYONaajFvHSD3-fra45iMRW3BQ9OG2gF1GmLH44me2nzZbcta4PWUAAvWu93t930LJYXVakda032ZMdoCfJlNH6XM7+y2t6ZRiHKLgyz0UtpjuCtJnAnFIO9SRFGUD9Pz9PVA34YCy2ZKceH7FCUyXdCgJAiDmGZQi8PwuB1xjOMExgJNU07PdMwPPRiKwmjN3o3tnxTJ5DD4h1YjzflfVeIA